### DISCUSSION TITLE
Internal(KvmContext(KvmCap(Xcrs))) on an old server

### DISCUSSION THREAD

**Discussion by bcspragu** [firecracker-microvm/firecracker#3470] [Q&A]
Hi there!

I'm trying to run Firecracker VMs on an old (okay, _really_ old) bare metal blade server. I've enabled virtualization in the BIOS, and have set up `/dev/kvm` as instructed, but I can't get VMs to run, they fail with the following:

```bash
$ firecracker --no-api --config-file ~/vmconfig.json
2023-02-22T19:06:20.817652022 [anonymous-instance:main:ERROR:src/firecracker/src/main.rs:480] Building VMM configured from cmdline json failed: Internal(KvmContext(KvmCap(Xcrs)))
```

I'm not sure if the server is literally just too old and lacking capabilities to run Firecracker, but according to [the datasheet](https://ark.intel.com/content/www/us/en/ark/products/47922/intel-xeon-processor-x5650-12m-cache-2-66-ghz-6-40-gts-intel-qpi.html) it does support VT-x.

Some debugging info:

```bash
$ ./tools/devtool checkenv
[Firecracker devtool] Checking prerequisites for running Firecracker.
[Firecracker devtool] Please check ../docs/getting-started.md#prerequisites in case of any error.
[Firecracker devtool] Checking Host Security Configuration.
[Firecracker devtool] Please check ../docs/prod-host-setup.md in case of any error.
[Firecracker devtool] WARNING: retpoline, IBPB, IBRS: DISABLED.
[Firecracker devtool] WARNING: VMX: cache flushes: DISABLED
[Firecracker devtool] WARNING: Hyperthreading ENABLED.

$ uname -a
Linux localhost 5.15.89-flatcar #1 SMP Wed Feb 15 18:00:42 -00 2023 x86_64 Intel(R) Xeon(R) CPU X5650 @ 2.67GHz GenuineIntel GNU/Linux

$ firecracker --version
Firecracker v1.1.4

Supported snapshot data format versions: v0.23.0, v0.24.0, v0.25.0, v1.0.0, v1.1.0
```

Not sure if the above `WARNING`s are actually a problem, or really just warnings.

**Comment by zulinx86** [comment]
Hi @bcspragu ,

The error seems to occur due to the lack of a KVM capability.
Firecracker requires `KVM_CAP_XCRS` capability to call `KVM_GET_XCRS` and `KVM_SET_XCRS` [[1]][[2]][[3]][[4]].

And `KVM_CAP_XCRS` capability depends on available features of CPU that you're using [[5]][[6]].
Could you check if your CPU support XSAVE instructions with the following command?
```
$ lscpu | grep xsave
```

[1]: https://www.kernel.org/doc/html/latest/virt/kvm/api.html#kvm-get-xcrs
[2]: https://github.com/firecracker-microvm/firecracker/blob/v1.1.4/src/vmm/src/vstate/system.rs#L74
[3]: https://github.com/firecracker-microvm/firecracker/blob/v1.1.4/src/vmm/src/vstate/vcpu/x86_64.rs#L302
[4]: https://github.com/firecracker-microvm/firecracker/blob/v1.1.4/src/vmm/src/vstate/vcpu/x86_64.rs#L397
[5]: https://elixir.bootlin.com/linux/v5.15.89/source/arch/x86/kvm/x86.c#L4201
[6]: https://elixir.bootlin.com/linux/v5.15.89/source/arch/x86/include/asm/cpufeatures.h#L141

I hope this information is helpful for you.
Thanks,

**Reply by bcspragu** [reply]
Hi Takahiro,

Thanks for the info and the links! `$ lscpu | grep xsave` returns nothing, looks like I don't have the `KVM_CAP_XCRS` capability. Looking at the firecracker code, I can't tell what `xsave` and `xcrs` are actually used for, it looks like they're just propagated to some vCPU handler.

My next question would be, how essential are these to Firecracker's functionality? Is it something that could be turned into a [stern warning like `tsc_khz`](https://github.com/firecracker-microvm/firecracker/blob/ad7163093159c3af3b08c10b048a7df8f4950af0/src/vmm/src/vstate/vcpu/x86_64.rs#L309), or is it core to Firecracker's functionality? I can patch Firecracker and give it a shot, but any context would be helpful.

Here's the full `lscpu` output, in case that's useful.

<details>
<summary>lscpu</summary>

```
# An aside: looks like there's a fair number of unpatched vulns on these CPUs, which isn't surprising given their age. This is less concerning to me because I'm running (mostly) trusted workloads in a (mostly) offline fashion.
$ lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         40 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  24
  On-line CPU(s) list:   0-23
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Xeon(R) CPU           X5650  @ 2.67GHz
    CPU family:          6
    Model:               44
    Thread(s) per core:  2
    Core(s) per socket:  6
    Socket(s):           2
    Stepping:            2
    BogoMIPS:            5319.75
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid
                         aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 popcnt aes lahf_lm pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid dtherm ida arat flush_l1d
Virtualization features:
  Virtualization:        VT-x
Caches (sum of all):
  L1d:                   384 KiB (12 instances)
  L1i:                   384 KiB (12 instances)
  L2:                    3 MiB (12 instances)
  L3:                    24 MiB (2 instances)
NUMA:
  NUMA node(s):          2
  NUMA node0 CPU(s):     0,2,4,6,8,10,12,14,16,18,20,22
  NUMA node1 CPU(s):     1,3,5,7,9,11,13,15,17,19,21,23
Vulnerabilities:
  Itlb multihit:         KVM: Mitigation: VMX disabled
  L1tf:                  Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
  Mds:                   Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
  Meltdown:              Mitigation; PTI
  Mmio stale data:       Unknown: No mitigations
  Retbleed:              Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:            Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
  Srbds:                 Not affected
  Tsx async abort:       Not affected
```
</details>

**Reply by zulinx86** [reply]
Hi @bcspragu ,

Thanks for your questions. From what I've seen, they are required when taking/restoring snapshots.
It might be able to do the capability check when taking/restoring snapshots and remove it from the new microVM creation process. This makes your CPU able to launch microVM and come to get the error only when taking/restoring snapshots.
We'll discuss this topic in the next few weeks and get back to you.

Thanks,

**Reply by bcspragu** [reply]
Ah that would be awesome, thank you! Indeed, I don't need snapshot/restore functionality, I'd just like to run one-shot VMs

**Reply by zulinx86** [reply]
Thank you for waiting.

As a result of the discussion, we concluded that it is not a problem to do the snapshot feature-specific capability check just before using snapshot functionalities unless it does not affect performance. Until this improvement has arrived, I'd appreciate it if you could consider to test firecracker by removing the capability check temporarily and building from source.

Thanks,

**Reply by bcspragu** [reply]
Glad to hear it! I'll try removing the capability check and seeing what happens, will do that in the next few days

**Reply by bcspragu** [reply]
Happy to confirm I'm able to run a standard Linux microVM with [this hacked together patch applied](https://github.com/firecracker-microvm/firecracker/compare/main...bcspragu:firecracker:brandon/no-xsave?expand=1), running with the standard `sudo ./firecracker --no-api --config-file ~/vmconfig.example.json`

**Reply by zulinx86** [reply]
Thanks for your confirmation! Sounds great!

I haven't got time to do yet and we need to sort out what capabilities are used only in snapshot feature first. I hope you will continue to use your nice hacky patch if you use firecracker for test purpose.

Thanks,

**Reply by travisghansen** [reply]
I applied the patch on top of the weaveworks macvtap fork and am using it successfully now with flintlock. Similar situation, older amd hardware that was previously crashing with the error mentioned above.

**Reply by mattiaforc** [reply]
@travisghansen I've been trying to make it work with flintlock too - could you provide some guidance on how to apply the patch? Did you just rebuild it from source and replace the firecracker binary?

**Reply by travisghansen** [reply]
Yeah, simple as that.

**Reply by mattiaforc** [reply]
> Yeah, simple as that.

Thanks, it worked just fine!

**Comment by zulinx86** [comment]
PR #3967 added an ability to add/remove KVM capabilities to be checked with custom CPU templates. This feature will be available in the next minor release (v1.5.0). Regarding how to use the custom CPU template feature, please refer to [the public doc](https://github.com/firecracker-microvm/firecracker/blob/main/docs/cpu_templates/cpu-templates.md). Closing this discussion.
